operator/concurrentwatcher.go

```diff
@@ -0,0 +1,157 @@
+package operator
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana-app-sdk/resource"
+
+	"github.com/cespare/xxhash/v2"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+var (
+	initialBufferSize = 1024
+)
+
+type eventInfo struct {
+	ctx    context.Context
+	action ResourceAction
+	target resource.Object
+	source resource.Object
+}
+
+// ConcurrentWatcher is a struct that implements ResourceWatcher, but takes no action on its own.
+// For each method in (Add, Update, Delete) the event is added in a buffered queue and the corresponding
+// methods of the underlying ResourceWatcher are called concurrently.
+// The events are processed by a fixed number of workers running concurrently. Each worker processing
+// the events one-by-one. The events are sharded using hash mod algorithm, ensuring events for same object
+// end up in the same worker. This is to uphold the guarantee of in-order delivery of events for an object.
+type ConcurrentWatcher struct {
```

Comment:
A thought when looking this over again--this might be better as a wrapper for an informer, rather than a watcher, as it requires being `Run` like an informer (watchers tend to be passive), and we use it to wrap an informer's event handler function anyway when adding on the informer side of the `InformerController`. If we revert the `Informer` interface to still take a `resource.Object`, we can still hashMod the same way as well. Since logically, this is added on the informer-add end of things, rather than the watcher-add (like with wrapping a watcher with `OpinionatedWatcher`), I'm curious if naming this as an informer (and adhering to the `Informer` interface) would be a better logical fit for it. Thoughts?

Reply:
I totally agree. It fits more as an Informer wrapper, especially with the need to manage the life-cycle for underlying workers.

Reply:
I have updated the code to not leak the concurrency related logic into the `InformerController`, and instead the choice of abstraction is now a new `ConcurrentInformer` that wraps and existing informer.